wingtips-core/src/main/java/com/nike/wingtips/Tracer.java

```diff
@@ -517,6 +543,98 @@ public void completeSubSpan() {
     }
 
     /**
+     * Handles the implementation of {@link Span#close()} (for {@link AutoCloseable}) for spans to allow them to be
+     * used in try-with-resources statements. We do the work here instead of in {@link Span#close()} itself since we
+     * have more visibility into whether a span is valid to be closed or not given the current state of the span stack.
+     * <ul>
+     *     <li>
+     *         If the given span is already completed ({@link Span#isCompleted()} returns true) then an error will be
+     *         logged and nothing will be done.
+     *     </li>
+     *     <li>
+     *         If the span is the current span ({@link #getCurrentSpan()} equals the given span), then {@link
+     *         #completeRequestSpan()} or {@link #completeSubSpan()} will be called, whichever is appropriate.
+     *     </li>
+     *     <li>
+     *         If the span is *not* the current span ({@link #getCurrentSpan()} does not equal the given span), then
+     *         this may or may not be an error depending on whether the given span is managed by {@link Tracer} or not.
+     *         <ul>
+     *             <li>
+     *                 If the span is managed by us (i.e. it is contained in the span stack somewhere even though it's
+     *                 not the current span) then this is a wingtips usage error - the span should not be completed
+     *                 yet - and an error will be logged and the given span will be completed and logged to the
+     *                 "invalid span logger".
+     *             </li>
+     *             <li>
+     *                 Otherwise the span is not managed by us, and since there may be valid use cases for manually
+     *                 managing spans we must assume the call was intentional. No error will be logged, and the span
+     *                 will be completed and logged to the "valid span logger".
+     *             </li>
+     *             <li>
+     *                 In either case, the current span stack and MDC info will be left untouched if the given span
+     *                 is not the current span.
+     *             </li>
+     *         </ul>
+     *     </li>
+     * </ul>
+     *
+     * <p>NOTE: This is intentionally package-scoped. Only {@link Span#close()} should ever call this method.
+     */
+    void handleSpanCloseMethod(Span span) {
+        // See if this span has already been completed - if so then this method should not have been called.
+        if (span.isCompleted()) {
+            classLogger.error(
+                "WINGTIPS USAGE ERROR - An attempt was made to close() a span that was already completed. "
+                + "This call to Span.close() will be ignored. "
+                + "wingtips_usage_error=true, already_completed_span=true, trace_id={}, span_id={}",
+                span.getTraceId(), span.getSpanId(), new Exception("Stack trace for debugging purposes")
+            );
+            return;
+        }
+
```

Comment:
It's really nice you're providing this functionality.  I think users will have less of a need to delve into the internals of Span.

Perhaps you could factor these if's and else's out to a public getType() function that could report the type of the Span.  The type could be an enum that could be CURRENT_ROOT_SPAN, CURRENT_SUB_SPAN, NON_CURRENT_SPAN, UNMANAGED_SPAN, and then you could switch-case on the enum here.

Reply:
Great suggestion! I'll do that.